rclcpp_lifecycle/test/test_lifecycle_service_client.cpp

```diff
@@ -0,0 +1,346 @@
+    }
+
+    if (future_result.get()) {
+      return future_result.get()->available_states;
+    }
+
+    return std::vector<lifecycle_msgs::msg::State>();
+  }
+
+  std::vector<lifecycle_msgs::msg::TransitionDescription>
+  get_available_transitions(std::chrono::seconds time_out = 1s)
+  {
+    auto request = std::make_shared<lifecycle_msgs::srv::GetAvailableTransitions::Request>();
+
+    if (!client_get_available_transitions_->wait_for_service(time_out)) {
+      return std::vector<lifecycle_msgs::msg::TransitionDescription>();
+    }
+
+    auto future_result = client_get_available_transitions_->async_send_request(request);
+    auto future_status = future_result.wait_for(time_out);
+
+    if (future_status != std::future_status::ready) {
+      return std::vector<lifecycle_msgs::msg::TransitionDescription>();
+    }
+
+    if (future_result.get()) {
+      return future_result.get()->available_transitions;
+    }
+
+    return std::vector<lifecycle_msgs::msg::TransitionDescription>();
+  }
+
+  std::vector<lifecycle_msgs::msg::TransitionDescription>
+  get_transition_graph(std::chrono::seconds time_out = 1s)
+  {
+    auto request = std::make_shared<lifecycle_msgs::srv::GetAvailableTransitions::Request>();
+
+    if (!client_get_transition_graph_->wait_for_service(time_out)) {
+      return std::vector<lifecycle_msgs::msg::TransitionDescription>();
+    }
+
+    auto future_result = client_get_transition_graph_->async_send_request(request);
+    auto future_status = future_result.wait_for(time_out);
+
+    if (future_status != std::future_status::ready) {
+      return std::vector<lifecycle_msgs::msg::TransitionDescription>();
+    }
+
+    if (future_result.get()) {
+      return future_result.get()->available_transitions;
+    }
+
+    return std::vector<lifecycle_msgs::msg::TransitionDescription>();
+  }
+
+private:
+  std::shared_ptr<rclcpp::Client<lifecycle_msgs::srv::GetAvailableStates>>
+  client_get_available_states_;
+  std::shared_ptr<rclcpp::Client<lifecycle_msgs::srv::GetAvailableTransitions>>
+  client_get_available_transitions_;
+  std::shared_ptr<rclcpp::Client<lifecycle_msgs::srv::GetAvailableTransitions>>
+  client_get_transition_graph_;
+  std::shared_ptr<rclcpp::Client<lifecycle_msgs::srv::GetState>> client_get_state_;
+  std::shared_ptr<rclcpp::Client<lifecycle_msgs::srv::ChangeState>> client_change_state_;
+};
+
+
+class TestLifecycleServiceClient : public ::testing::Test
+{
+protected:
+  EmptyLifecycleNode * lifecycle_node() {return lifecycle_node_.get();}
+  LifecycleServiceClient * lifecycle_client() {return lifecycle_client_.get();}
+
+private:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+    lifecycle_node_ = std::make_shared<EmptyLifecycleNode>();
+    lifecycle_client_ = std::make_shared<LifecycleServiceClient>("client");
+    spinner_ = std::thread(&TestLifecycleServiceClient::spin, this);
+  }
+
+  void TearDown() override
+  {
+    rclcpp::shutdown();
```

Comment:
Calling `shutdown` from a different thread causes a race which was making this test fail in the nightly jobs since it has been added. See #1204 for the proposed fix.